// database/User.ts
import { Model, DataTypes, Sequelize } from 'sequelize';

export async function UserFactory(sequelize: Sequelize, config: { SYNC_SEQUELIZE_MODELS: string }) {
    class User extends Model {
        public readonly id!: number; // Note that the `null assertion` `!` is required in strict mode.
        public readonly name!: string;
        public readonly email!: string;

        // timestamps!
        public readonly createdAt!: Date;
        public readonly updatedAt!: Date;

        public readonly facebook: string;
    }

    User.init({
        id: {
            type: DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        name: {
            type: new DataTypes.STRING(128),
            allowNull: false,
        },
        email: {
            type: new DataTypes.STRING(128),
            allowNull: false,
        },
        facebook: {
            type: new DataTypes.STRING(10000),
            allowNull: true,
        }
    }, {
            tableName: 'User',
            sequelize
        }
    );

    if (config.SYNC_SEQUELIZE_MODELS) {
        await User.sync({ force: true })
    }

    return User;
}

type Unpromise<T extends Promise<any>> = T extends Promise<infer U> ? U : never;
export type UserType = Unpromise<ReturnType<typeof UserFactory>>
